### DISCUSSION TITLE
Determining if SQL string contains no statement

### DISCUSSION THREAD

**Discussion by mlabrec** [JSQLParser/JSqlParser#1120] [Q&A]
I have observed that JsqlParser throws an error when trying to parse an empty query. For example, the following queries all throw an error:

CCJSqlParserUtil.parse("");
CCJSqlParserUtil.parse("    ");
CCJSqlParserUtil.parse(" -- some comment\n /* other comment */  ");

The error thrown is ambiguous, and is either "Encountered unexpected token:<EOF>" or "Array index out of range: -1".

For my use case, users enter an SQL string into a text field, and JsqlParser is used to parse the string. I would like to specifically determine the case where the string is empty and/or only comments. Checking for empty or whitespace only strings is easy to do outside of JsqlParser. But, it is not straightforward to determine if the string is empty once filtering out comments. Is there any way to determine this with JsqlParser?

**Comment by wumpz** [comment]
Interesting use case. I guess we need a new starting grammar rule for this.

**Reply by manticore-projects** [reply]
For JSQLFormatter I have solved this very problem, but using a Java based Pre-Processor module:

1) Identify all kind SQL comments with Start and End Position relative to the actual SQL content (ignoring any kind of white space OUTSIDE of a comment).
2) When the length of the normalized SQL content (after removing all white space) equals the TOTAL length of the comments, then there are no actual SQL statements

Hope this helps.

**Reply by wumpz** [reply]
So what would be the expected outcome of this? parse("") == null?

**Reply by manticore-projects** [reply]
> So what would be the expected outcome of this? parse("") == null?

I see 2 possibilities:
1) throw a specific `NO ACTUAL STATEMENT` Exception or
2) return `NULL` (which reflects that there was no actual statement)

I understand, that there is already a `ParseException` thrown, but this is very unspecific.
See it from the perspective of a Code Formatting Editor, which catches the exception and then needs to decide what to do, e.g. where to point to for the errors. If there is no actual statement, then this Code Formatting Editor must not do anything, while he must react in case of parse exceptions for actual statements.

**Reply by manticore-projects** [reply]
There is one more option of course: Define a new `VoidStatement` which is just an empty class without any properties and reflects an "empty SQL statement".
